pkg/antctl/raw/multicluster/join.go

```diff
@@ -0,0 +1,314 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package multicluster
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	yamlutil "k8s.io/apimachinery/pkg/util/yaml"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	multiclusterv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+	"antrea.io/antrea/pkg/antctl/raw/multicluster/common"
+)
+
+const (
+	defaultMemberNamespace = "kube-system"
+)
+
+type ClusterSetJoinConfig struct {
+	Kind            string     `yaml:"kind"`
+	APIVersion      string     `yaml:"apiVersion"`
+	ClusterSetID    string     `yaml:"clusterSetID"`
+	ClusterID       string     `yaml:"clusterID"`
+	Namespace       string     `yaml:"namespace,omitempty"`
+	LeaderClusterID string     `yaml:"leaderClusterID"`
+	LeaderNamespace string     `yaml:"leaderNamespace"`
+	LeaderAPIServer string     `yaml:"leaderAPIServer"`
+	TokenSecretName string     `yaml:"tokenSecretName,omitempty"`
+	TokenSecretFile string     `yaml:"tokenSecretFile,omitempty"`
+	ConfigFile      string     `yaml:"-"`
+	Secret          *v1.Secret `yaml:"-"`
+}
+
+var joinOpts *ClusterSetJoinConfig
+
+func (o *ClusterSetJoinConfig) validateAndComplete() error {
+	if o.ConfigFile != "" {
+		raw, err := ioutil.ReadFile(o.ConfigFile)
+		if err != nil {
+			return err
+		}
+		if err := yamlUnmarshall(raw, o); err != nil {
+			return err
+		}
+		if o.Kind != common.ClusterSetJoinConfigKind || o.APIVersion != common.ClusterSetJoinConfigAPIVersion {
+			return fmt.Errorf("unknown apiVersion or kind in config: %s", o.ConfigFile)
+		}
+
+		if o.TokenSecretName == "" && o.TokenSecretFile == "" {
+			o.Secret, err = unmarshallSecret(raw, true)
+			if err != nil {
+				return fmt.Errorf("failed to unmarshall Secret from config file: %v", err)
+			}
+			o.Secret.Namespace = o.Namespace
```

Comment:
This line can be removed (already covered by line 119.

Reply:
Removed.